pkg/tests/schema_pulumi_test.go

```diff
@@ -4159,3 +4012,3298 @@ func TestMakeTerraformResultNilVsEmptyMap(t *testing.T) {
+	pt.Up(t)
+	pulumiYamlPath := filepath.Join(pt.CurrentStack().Workspace().WorkDir(), "Pulumi.yaml")
+
+	err := os.WriteFile(pulumiYamlPath, []byte(program2), 0o600)
+	require.NoError(t, err)
+
+	pt.ClearGrpcLog(t)
+	res := pt.Preview(t, optpreview.Diff())
+	t.Log(res.StdOut)
+
+	diffResponse := struct {
+		DetailedDiff map[string]interface{} `json:"detailedDiff"`
+	}{}
+
+	for _, entry := range pt.GrpcLog(t).Entries {
+		if entry.Method == "/pulumirpc.ResourceProvider/Diff" {
+			err := json.Unmarshal(entry.Response, &diffResponse)
+			require.NoError(t, err)
+		}
+	}
+
+	return res.StdOut, diffResponse.DetailedDiff
+}
+
+func TestDetailedDiffSet(t *testing.T) {
+	// TODO: Remove this once accurate bridge previews are rolled out
+	t.Setenv("PULUMI_TF_BRIDGE_ACCURATE_BRIDGE_PREVIEW", "true")
+	runTest := func(t *testing.T, resMap map[string]*schema.Resource, props1, props2 interface{},
+		expected autogold.Value, expectedDetailedDiff map[string]any,
+	) {
+		program := `
+name: test
+runtime: yaml
+resources:
+  mainRes:
+    type: prov:index:Test
+    properties:
+      tests: %s
+`
+		props1JSON, err := json.Marshal(props1)
+		require.NoError(t, err)
+		program1 := fmt.Sprintf(program, string(props1JSON))
+		props2JSON, err := json.Marshal(props2)
+		require.NoError(t, err)
+		program2 := fmt.Sprintf(program, string(props2JSON))
+		out, detailedDiff := runDetailedDiffTest(t, resMap, program1, program2)
+
+		expected.Equal(t, out)
+		require.Equal(t, expectedDetailedDiff, detailedDiff)
+	}
+
+	type setDetailedDiffTestCase struct {
+		name                              string
+		props1                            []string
+		props2                            []string
+		expectedAttrDetailedDiff          map[string]any
+		expectedAttr                      autogold.Value
+		expectedAttrForceNewDetailedDiff  map[string]any
+		expectedAttrForceNew              autogold.Value
+		expectedBlockDetailedDiff         map[string]any
+		expectedBlock                     autogold.Value
+		expectedBlockForceNewDetailedDiff map[string]any
+		expectedBlockForceNew             autogold.Value
+	}
+
+	testCases := []setDetailedDiffTestCase{
+		{
+			"unchanged",
+			[]string{"val1"},
+			[]string{"val1"},
+			nil,
+			autogold.Expect(`Previewing update (test):
+  pulumi:pulumi:Stack: (same)
+    [urn=urn:pulumi:test::test::pulumi:pulumi:Stack::test-test]
+Resources:
+    2 unchanged
+`),
+			nil,
+			autogold.Expect(`Previewing update (test):
+  pulumi:pulumi:Stack: (same)
+    [urn=urn:pulumi:test::test::pulumi:pulumi:Stack::test-test]
+Resources:
+    2 unchanged
+`),
+			nil,
+			autogold.Expect(`Previewing update (test):
+  pulumi:pulumi:Stack: (same)
+    [urn=urn:pulumi:test::test::pulumi:pulumi:Stack::test-test]
+Resources:
+    2 unchanged
+`),
+			nil,
+			autogold.Expect(`Previewing update (test):
+  pulumi:pulumi:Stack: (same)
+    [urn=urn:pulumi:test::test::pulumi:pulumi:Stack::test-test]
+Resources:
+    2 unchanged
+`),
+		},
```

Comment:
I don't understand how these test cases work. Can we (0) add comments explaining what the different fields mean, (1) use labels on the inputs and (2) move the autogold output to be file based.
```suggestion
		{
			name: "unchanged",
			beforeState: []string{"val1"},
			afterState: []string{"val1"},
		},
```

(1) would allow us to omit `nil` values and (2) would make it possible to scan the test lists.

Reply:
Same here, these are harder to read then the above test cases that are excellent. I actually like inline "Previewing update" verbiage unlike Ian, but I would like to be able to more easily tell when reading a single test case without too much extra context, that what I am seeing in the autogold.Expect() is indeed what we want to expect.

Reply:
Moving the outputs to a separate file would make this completely unmaintainable. I've instead trimmed the recorded outputs so that only the essential part of the diff is saved in the file - that should make the test cases a lot more readable.